src/source/nats.rs

```diff
@@ -0,0 +1,178 @@
+pub struct Config {
+    // list of hosts
+    pub hosts: Vec<String>,
+    // subject to send messages to
+    pub subject: String,
+    // optional queue to subscribe to
+    #[serde(default = "Default::default")]
+    pub queue: Option<String>,
+    // options to use when opening a new connection
+    #[serde(default = "Default::default")]
+    pub options: ConnectOptions,
+}
+
+impl ConfigImpl for Config {}
+
+impl Config {
+    async fn connection(&self) -> Result<NatsConnection> {
+        let hosts = self.hosts.join(",");
+        let connection = self.options.generate()?.connect(hosts.as_str()).await?;
+        Ok(connection)
+    }
+}
+
+pub struct Nats {
+    pub config: Config,
+    onramp_id: TremorUrl,
+}
+
+impl onramp::Impl for Nats {
+    fn from_config(id: &TremorUrl, config: &Option<YamlValue>) -> Result<Box<dyn Onramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            Ok(Box::new(Self {
+                config,
+                onramp_id: id.clone(),
+            }))
+        } else {
+            Err("Missing config for nats onramp".into())
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Onramp for Nats {
+    async fn start(&mut self, config: OnrampConfig<'_>) -> Result<onramp::Addr> {
+        let source = Int::from_config(config.onramp_uid, self.onramp_id.clone(), &self.config);
+        SourceManager::start(source, config).await
+    }
+
+    fn default_codec(&self) -> &str {
+        "json"
+    }
+}
+
+pub struct Int {
+    onramp_id: TremorUrl,
+    config: Config,
+    subscription: Option<Subscription>,
+    connection: Option<NatsConnection>,
+    origin_uri: EventOriginUri,
+}
+
+impl std::fmt::Debug for Int {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NATS")
+    }
+}
+
+impl Int {
+    fn from_config(uid: u64, onramp_id: TremorUrl, config: &Config) -> Self {
+        let config = config.clone();
+        let origin_uri = EventOriginUri {
+            uid,
+            scheme: "tremor-nats".to_string(),
+            host: "not-connected".to_string(),
+            port: None,
+            path: vec![],
+        };
+        Self {
+            onramp_id,
+            config,
+            subscription: None,
+            connection: None,
+            origin_uri,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for Int {
+    async fn pull_event(&mut self, _id: u64) -> Result<SourceReply> {
+        if let Some(sub) = &self.subscription {
+            if let Some(msg) = sub.next().await {
+                let mut origin_uri = self.origin_uri.clone();
+                origin_uri.path = vec![msg.subject];
+                let data = msg.data;
+                let mut nats_meta_data = Value::object_with_capacity(1);
+                let mut msg_reply = None;
+                if let Some(reply) = msg.reply {
+                    msg_reply = Some(reply);
+                }
```

Comment:
The more rusty way is using an expression instead of mutating, it generally makes the code easier to reason about:
```rust
                let msg_reply = if let Some(reply) = msg.reply {
                    Some(reply)
                } else {
                    None
                }
```

it also maps it easier to spot that this basically is:
```suggestion
                let msg_reply = msg.reply;
```